### PR TITLE
Imprv/166579 166580 improve github diff position

### PR DESCRIPTION
**Pull request by atsuki-t**
https://redmine.weseek.co.jp/issues/166607

## やったこと
- [ram-lab][code-hedgehog] コメントの position とファイルの行数を一致させる
  - Github のコメント作成 API で使用するプロパティを変更 (position -> line)
  - playground で正確な github web ui の diff に飛べるようにした

## スクリーンショット
- コメントカードから適切な位置の差分行に飛べることを確認
image

### CODE REVIEW

File: packages/processors/base/processor.ts [hunk 1]
```diff
@@ -305,9 +305,40 @@ export abstract class BaseProcessor implements IPullRequestProcessor {
...
+          numberedLines.push(line);
+        }
+        oldLineNumber++;
+        newLineNumber++;
+      } else {
+        // Other lines (file headers, etc.)
+        numberedLines.push(line);
+      }
+    }
+
```
Comment (atsuki-t) [inline packages/processors/base/processor.ts:341]: ## 修正前の addLineNumbersToDiff メソッドの返り値
- position の値 ("@@" ヘッダーから何行目か) の情報を付与して返していた
```
0: @@ -137,7 +139,9 @@ const PullRequestContent = ({ pullRequest, githubToken, owner, repo, number }: P
1:          )}
2:          <div>
3:            <h1 className="text-xl font-semibold mb-2">
4: -            {pullRequest.title} <span className="text-muted-foreground font-normal">#{pullRequest.number}</span>
5: +            <a href={'https://github.com'} className="hover:underline" target="_blank" rel="noopener noreferrer">
6: +              {pullRequest.title} <span className="text-muted-foreground font-normal">#{pullRequest.number}</span>
7: +            </a>
8:            </h1>
9:            <div className="flex flex-wrap gap-x-4 gap-y-2 text-sm text-muted-foreground">
10:              <Badge variant={state} className="flex items-center gap-1">
```

## 修正後の addLineNumbersToDiff メソッドの返り値
- line の値を付与して返すようにした
```
@@ -137,7 +139,9 @@ const PullRequestContent = ({ pullRequest, githubToken, owner, repo, number }: P
139:          )}
140:          <div>
141:            <h1 className="text-xl font-semibold mb-2">
140: -            {pullRequest.title} <span className="text-muted-foreground font-normal">#{pullRequest.number}</span>
142: +            <a href={'https://github.com'} className="hover:underline" target="_blank" rel="noopener noreferrer">
143: +              {pullRequest.title} <span className="text-muted-foreground font-normal">#{pullRequest.number}</span>
144: +            </a>
145:            </h1>
146:            <div className="flex flex-wrap gap-x-4 gap-y-2 text-sm text-muted-foreground">
147:              <Badge variant={state} className="flex items-center gap-1">
```

File: packages/playground/src/components/PullRequestDetail.tsx
Comment (atsuki-t) [inline packages/playground/src/components/PullRequestDetail.tsx:1]: playground で指摘箇所に飛べるようにした

**Comment by atsuki-t** [comment]
## Overall Summary

This update further enhances the system's handling of GitHub diff positions and comment mapping, building on previous improvements to align more closely with GitHub's API and UI behavior. In `PullRequestDetail.tsx`, the display logic for comment line positions has been revised: the display text is simplified from '差分内で上から〇行目の位置' (the Nth line from the top of the diff) to just '〇行' (N line), directly reflecting the actual line number in the diff. Additionally, the generated GitHub file diff links now include the precise line anchor using the format `R{position}`, ensuring that users are navigated directly to the exact line referenced in the comment when following the link. In `dify/processor.ts`, the assignment of the `patch` property is updated to use `addLineNumbersToDiff`, enriching the diff with accurate line number annotations. This supports improved precision when specifying comment positions via the GitHub API, which now relies on the `line` property rather than `position`. Together, these changes strengthen the consistency and clarity of comment placement, both in the UI and when interfacing with GitHub's backend, ensuring seamless navigation and context for users and integrators.

## Reviewed Changes

| File | Description |
|------|-------------|
| `packages/core/src/vcs/github.ts` | The code change correctly updates the GitHub API payload to use the modern 'line' and 'side' parameters for comment placement, which should improve comment accuracy and user navigation in most scenarios. However, the update assumes the upstream mapping is already correct and does not account for left-side commenting or complex diff scenarios. There are no obvious performance concerns introduced by this change, as it only modifies the payload structure. To ensure robustness and maintainability, review and test the diff mapping logic thoroughly, update all related code paths and documentation, and consider defensive error handling for API failures. Overall, this is a good incremental improvement, but attention to edge cases and holistic updates across the codebase are recommended. |
| `packages/playground/src/components/PullRequestDetail.tsx` | The code changes improve UX by providing direct navigation to specific diff lines in GitHub and simplifying the display of line numbers. Edge case handling and user clarity could be improved by validating inputs and adding contextual help. There are no performance or security regressions introduced by this patch. Overall, the update is a positive step for usability and consistency, but minor enhancements for robustness and clarity are recommended. |
| `packages/processors/base/processor.ts` | The new diff line-numbering logic accurately tracks old and new file line numbers and annotates them per diff line, closely matching GitHub's expectations for comment positioning. This is a clear improvement for accuracy and user experience. There are minor concerns about error handling for malformed diffs and some edge cases, as well as opportunities for improving documentation and maintainability. No major performance or security issues are apparent. With a few small tweaks and expanded test coverage, this code will be robust and production-ready. |
| `packages/processors/dify/processor.ts` | The update to process and annotate patch diffs with line numbers enhances accuracy for comment navigation and API integration, aligning backend processing with improved UI and API workflows. However, this introduces potential performance overhead and error risks, especially when dealing with large or malformed patch data. Key recommendations are to validate inputs rigorously, optimize 'addLineNumbersToDiff' for large diffs, and ensure robust error handling. The architectural change is positive for maintainability and clarity, but careful testing and documentation are required to prevent regressions or efficiency issues in production. |



<details>
<summary>Comments suppressed due to low severity (6)</summary>

**packages/core/src/vcs/github.ts**
- Setting 'side' statically to 'RIGHT' means comments can only be placed on the 'new' (post-change) side of the diff. This is correct for most review workflows but does not support legacy or edge cases where left-side (base/original) comments are needed.
  - If only supporting comments on the new side is a business requirement, document this limitation. Otherwise, add logic to determine the appropriate 'side' based on the comment context.
- No additional error handling is introduced for API failures or mapping errors. If the mapping is incorrect or the API rejects the payload, comments may be silently lost or misplaced.
  - Add error handling for API responses and log or surface errors if the comment cannot be placed as expected.


**packages/playground/src/components/PullRequestDetail.tsx**
- No memoization or caching is present around the construction of the GitHub diff link, but as this is a UI rendering small string, it is not a performance concern in the current context.
- No excessive data retrieval or unnecessary computation is present in this change. The modification is lightweight and should not noticeably impact performance.


**packages/playground/src/components/PullRequestDetail.tsx**:205
- The display of the line position has been simplified from a verbose string to just '{comment.position ?? '-'} 行'. While this makes the UI more concise, it may reduce clarity for users unfamiliar with the context (e.g., is this the line number in the file, the diff, or something else?).
  - Consider adding a tooltip or help icon to clarify what the displayed line number refers to (e.g., 'この行番号はGitHubの差分表示に対応しています' or similar).


**packages/playground/src/components/PullRequestDetail.tsx**:207
- The code assumes that `comment.diffId` and `comment.position` are always available and valid for constructing the anchor. If these fields can be missing or malformed in some data, it could lead to broken links or runtime errors.
  - Add defensive checks or error boundaries to handle unexpected null/undefined values for `comment.diffId` or `comment.position`, possibly displaying a warning in the UI if data is incomplete.


**packages/processors/base/processor.ts**
- Parsing of hunk headers with regex assumes both old and new line numbers are always present. Some diffs may omit the count (e.g., '@@ -1 +1 @@'), which still matches, but the current regex does not handle malformed hunks gracefully.
  - Add error handling for malformed hunk headers, and consider logging or throwing an error if the match fails to prevent silent bugs.
- The logic increments both oldLineNumber and newLineNumber for empty lines. While this mirrors the original, it could be incorrect if the diff contains leading/trailing blank lines not corresponding to the source file. This could potentially skew line number alignment in rare cases.
  - Add inline documentation and a comment clarifying the intent, or add a check to skip blank lines at the start/end of the diff if they don't correspond to actual file content.
- Performance: The function iterates over the diff lines only once, which is efficient. However, returning a fully joined string could be wasteful if only partial patch information is needed later (e.g., for large diffs).
  - If use cases arise where line-by-line access is required, consider returning the array with number annotations and joining only when necessary.
- Readability: The variable names and logic are clear, but a top-level comment describing the numbering strategy and its mapping to GitHub's 'line' property would improve maintainability, especially for future readers unfamiliar with diff format details.
  - Add a block comment at the top of the function summarizing the numbering approach and its relation to the GitHub API.


**packages/processors/dify/processor.ts**
- If 'patch' is a large string, adding line numbers may significantly increase memory usage and output size, especially if the patch is already large. This could impact performance and resource efficiency, particularly in bulk processing.
  - If only certain consumers require line-numbered diffs, consider adding this processing conditionally. Alternatively, provide an option to skip or defer this computation unless strictly necessary.
- The code change improves maintainability by centralizing and encapsulating the diff line-numbering logic in a dedicated function, which is good. However, ensure that this function is well-documented and tested, as it is now a critical part of the comment mapping pipeline.
  - Add or update documentation and unit tests for 'addLineNumbersToDiff', especially for edge cases such as multi-hunk diffs, binary files, or patches with unusual formatting.
</details>

File: packages/processors/base/processor.ts (hunk 1 above)
Comment (yuki-takei) [reply packages/processors/base/processor.ts:341]: あれこれ結構前に議論しなかったっけ🙄

ユーティリティ化してほしいです

**Review by yuki-takei — CHANGES_REQUESTED** [review]
